scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/utils.js

```diff
@@ -71,28 +71,22 @@ function isMacro(node, macroName) {
   );
 }
 
-/**
- * Log a message when a file is missing an ingredient.
- *
- * @param {VFile} file - a VFile
- * @param {Object} context - a context object with recipe name and ingredient
- * strings
- */
-function logMissingIngredient(file, context) {
-  const { recipeName, ingredient, source } = context;
-  const rule = `${recipeName}/${ingredient}`;
-  const origin = `${source}:${rule}`;
-
-  const message = file.message(
-    `Missing from ${recipeName}: ${ingredient}`,
-    origin
-  );
-  message.fatal = true;
+function Logger(file, source, recipeName, ingredient) {
+  return {
+    fail: function(node, text, id) {
```

Comment:
I like more consistency but I'm worried this would be too restrictive. For the class members I'm going to want some version of "list items must be alphabetised" and "list items must contain links", and I'm not sure how we would express these in a format like that. Wording is really hard for error messages :).

Reply:
Yeah, I understand that not all messages will be of this "expected _thing_" form. But all of the messages we currently have are of this form. We could rename this method to `Logger.expected(node, name, id)` method and have a separate `Logger.fail(node, text, id)` method for free-form messages (I would expect `expected` to call `fail` internally).

Reply:
I just tried this, and it is nice :). We might want to think about other ways to formalize messages as we add new forms, but this looks great for now.